src/content/developers/docs/consensus-mechanisms/pow/mining-algorithms/dagger-hashamoto/index.md

```diff
@@ -0,0 +1,327 @@
+
+```python
+def orig_hashimoto(prev_hash, merkle_root, list_of_transactions, nonce):
+    hash_output_A = sha256(prev_hash + merkle_root + nonce)
+    txid_mix = 0
+    for i in range(64):
+        shifted_A = hash_output_A >> i
+        transaction = shifted_A % len(list_of_transactions)
+        txid_mix ^= list_of_transactions[transaction] << i
+    return txid_max ^ (nonce << 192)
+```
+
+Unfortunately, while Hashimoto is considered RAM hard, it relies on 256-bit arithmetic, which has considerable computational overhead. To address this issue, dagger hashimoto only uses the least significant 64 bits when indexing its dataset.
+
+```python
+def hashimoto(dag, dagsize, params, header, nonce):
+    m = dagsize / 2
+    mix = sha3(encode_int(nonce) + header)
+    for _ in range(params["accesses"]):
+        mix ^= dag[m + (mix % 2**64) % m]
+    return dbl_sha3(mix)
+```
+
+The use of double sha3 allows for a form of zero-data, near-instant pre-verification, verifying only that a correct intermediate value was provided. This outer layer of PoW is highly ASIC-friendly and fairly weak, but exists to make DDoS even more difficult since that small amount of work must be done in order to produce a block that will not be rejected immediately. Here is the light-client version:
+
+```python
+def quick_hashimoto(seed, dagsize, params, header, nonce):
+    m = dagsize // 2
+    mix = sha3(nonce + header)
+    for _ in range(params["accesses"]):
+        mix ^= quick_calc(params, seed, m + (mix % 2**64) % m)
+    return dbl_sha3(mix)
+```
+
+## Mining and Verifying
+
+Now, let us put it all together into the mining algo:
+
+```python
+def mine(daggerset, params, block):
+    from random import randint
+    nonce = randint(0, 2**64)
+    while 1:
+        result = hashimoto(daggerset, get_dagsize(params, block),
+                           params, decode_int(block.prevhash), nonce)
+        if result * params["diff"] < 2**256:
+            break
+        nonce += 1
+        if nonce >= 2**64:
+            nonce = 0
+    return nonce
+```
+
+Here is the verification algorithm:
+
+```python
+def verify(daggerset, params, block, nonce):
+    result = hashimoto(daggerset, get_dagsize(params, block),
+                       params, decode_int(block.prevhash), nonce)
+    return result * params["diff"] < 2**256
+```
+
+Light-client friendly verification:
+
+```python
+def light_verify(params, header, nonce):
+    seedset = get_seedset(params, block)
+    result = quick_hashimoto(seedset["front_hash"], get_dagsize(params, block),
+                             params, decode_int(block.prevhash), nonce)
+    return result * params["diff"] < 2**256
+```
+
+Also, note that Dagger Hashimoto imposes additional requirements on the block header:
+
+- For two-layer verification to work, a block header must have both the nonce and the middle value pre-sha3
+- Somewhere, a block header must store the sha3 of the current seedset
+
+## Appendix
```

Comment:
Not a blocker but we could use `<fn>` tags to implement the appendix [similar to our NFT page](https://ethereum.org/en/nft/#fn-1)